### PR TITLE
Add network mapper from bitshares-1

### PR DESCRIPTION
**Pull request by pmconrad**
Fix for #1550 

Ported from https://github.com/bitshares/bitshares1-core/blob/master/programs/utils/map_bts_network.cpp

### CODE REVIEW

File: programs/network_mapper/network_mapper.cpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    fc::sha512 shared_secret = _connection->get_shared_secret();
+    shared_secret_encoder.write(shared_secret.data(), sizeof(shared_secret));
+    fc::ecc::compact_signature signature = my_node_id.sign_compact(shared_secret_encoder.result());
+
+    graphene::net::hello_message hello("network_mapper",
+                                  GRAPHENE_NET_PROTOCOL_VERSION,
+                                  fc::ip::address(), 0, 0,
+                                  my_node_id.get_public_key(),
+                                  signature,
+				  chain_id,
```
Comment (cogutvalera) [inline programs/network_mapper/network_mapper.cpp]: this line is looking ugly

File: programs/network_mapper/network_mapper.cpp [hunk 2]
```diff
@@ -173,45 +158,35 @@ class peer_probe : public graphene::net::peer_connection_delegate
...
-    if (colon_pos == std::string::npos)
-        FC_THROW("Missing required port number in endpoint string \"${endpoint_string}\"",
-                 ("endpoint_string", endpoint_string));
-    std::string port_string = endpoint_string.substr(colon_pos + 1);
-    uint16_t port = boost::lexical_cast<uint16_t>(port_string);
-    std::string hostname = endpoint_string.substr(0, colon_pos);
-    return fc::resolve(hostname, port);
-}
-
 int main(int argc, char** argv)
```
[the remark this answers is not in the document]
Comment (cogutvalera) [reply programs/network_mapper/network_mapper.cpp]: in addition IMHO we should have some default values

**Comment by abitmore** [comment]
@cogutvalera:
> boost::bad_any_cast

Please merge latest develop when testing. I believe it's fixed by https://github.com/bitshares/bitshares-core/pull/1374.

**Comment by cogutvalera** [comment]
@abitmore Thanks ! ok sure will bump now, I believe too that it will help, just thought about this too ...

**Comment by cogutvalera** [comment]
rebase helped !

**Comment by pmconrad** [comment]
Thanks. If whoever tackles #1348 decides that this tool might be handy, we can invest time to reformat and think about ``boost::program_options``.

**Comment by oxarbitrage** [comment]
I wanted to giuve a try to this tool however i am getting the following error with any seed i tried:

```
alfredo@alfredo-System-Product-Name:~/CLionProjects/pull1372/bitshares-core$ ./programs/network_mapper/network_mapper 4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8 bts-seed1.abit-more.com:62015
116.239.98.243:62015 ran into an error!
0 checked, 0 active, 0 to do
alfredo@alfredo-System-Product-Name:~/CLionProjects/pull1372/bitshares-core$ 
```

```
alfredo@alfredo-System-Product-Name:~/CLionProjects/pull1372/bitshares-core$ ./programs/network_mapper/network_mapper 4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8 95.179.137.33:1776
95.179.137.33:1776 ran into an error!
0 checked, 0 active, 0 to do
alfredo@alfredo-System-Product-Name:~/CLionProjects/pull1372/bitshares-core$
```

I am not running it properly ?

**Comment by pmconrad** [comment]
Hm, seems to be some kind of initialization problem. If you specify two seed nodes (or one with several addresses) it'll work.

**Comment by ryanRfox** [comment]
@pmconrad I modified the PR Description to reference #1550 so I believe it will close that Issue upon merge and leave the original Issue open for now.

**Comment by oxarbitrage** [comment]
I have a hard time understanding the usefulness of this program. Here is a sample log: https://raw.githubusercontent.com/oxarbitrage/logs/master/network-mapper.txt

Can someone direct me to a practical example on what someone can do with this output ?

Another thing, if we merge we need to update the programs README if we add this new program: https://github.com/bitshares/bitshares-core/blob/develop/programs/README.md

**Comment by dnotestein** [comment]
> I have a hard time understanding the usefulness of this program. Here is a sample log: https://raw.githubusercontent.com/oxarbitrage/logs/master/network-mapper.txt

It generates .DOT file that can be rendered by graphviz to make images of node connectivity.

The DOT file is created here:
  std::ofstream dot_stream((data_dir / "network_graph.dot").string().c_str());

**Comment by oxarbitrage** [comment]
Thank you @dnotestein , makes sense now.

However it seems to not be working, the dot file is not generated:

```
alfredo@alfredo-System-Product-Name:/tmp/network_map_4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8$ ls
alfredo@alfredo-System-Product-Name:/tmp/network_map_4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8$ 
```

maybe because of segfault?

```
Failed to connect 94.130.120.250:51174 - skipping!
2338483ms th_m       network_mapper.cpp:51         start                ] timeout connecting to node 104.237.1.18:1776
Failed to connect 104.237.1.18:1776 - skipping!
terminate called after throwing an instance of 'fc::assert_exception'

Thread 1 "network_mapper" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6ab5801 in __GI_abort () at abort.c:79
#2  0x00007ffff74a88fb in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff74aed3a in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff74aed95 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff74aefe8 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00005555556da203 in fc::ecc::public_key::serialize (this=<optimized out>) at /home/alfredo/CLionProjects/pull1372/bitshares-core/libraries/fc/src/crypto/elliptic_secp256k1.cpp:119
#7  0x0000555555632126 in fc::ecc::public_key::operator fc::array<char, 33ul> (this=<optimized out>)
    at /home/alfredo/CLionProjects/pull1372/bitshares-core/libraries/fc/include/fc/crypto/elliptic.hpp:42
#8  main (argc=<optimized out>, argv=<optimized out>) at /home/alfredo/CLionProjects/pull1372/bitshares-core/programs/network_mapper/network_mapper.cpp:243
```

**Comment by pmconrad** [comment]
Replaced with #1555, see comment there.

**Pull request closed without merging**